### ISSUE TITLE
zfs send a zpool residing on NVME devices makes the pool being unresponsive (device crash)

### ISSUE THREAD

**Issue by admnd** [openzfs/zfs#14770]
I throw a bottle at the sea, hoping to not be alone.

### System information

Gentoo Linux x86/64 with kernel 6.2.11  and ZFS 2.1.11
My system zpool is RAID-Z1 VDEV composed of 3x NVME WD SN770 modules with the latest available firmware applied and all configured with 4K sectors (No idea if a 512b sector size triggers the issue, this is my main machine). The motherboard is an Asus TUF Gaming X670E-Plus WiFi (upgraded to the latest available BIOS) having the modules are installed in the 3 PCIe 4.0 slots (moving a module in the PCIe 5.0 slot does not seem to change the situation). 

### Describe the problem you're observing

When I use  `zfs send` (the dataset is LZ4 compressed) one of my NVME modules crashes (sometimes the first, sometimes second or the third) and the pool is suspended so the transfer stalls forever.

The issue appears frequently on big data transfers (e.g. sending the whole system pool+its snapshots collection) to a remote TrueNAS box over  a local 10GbE network link. The transferred data size when the crash occurs randomly varies with no obvious pattern: sometimes the issue appears after 12Gb, sometimes after  78Gb, sometimes after 93 Gb and so on. Sometimes it completes the operation successfully.

When the crash happens, the following trace appears in the kernel logs:
```
[430771.216723] nvme nvme2: controller is down; will reset: CSTS=0xffffffff, PCI_STATUS=0x10
[430771.216727] nvme nvme2: Does your device have a faulty power saving mode enabled?
[430771.216729] nvme nvme2: Try "nvme_core.default_ps_max_latency_us=0 pcie_aspm=off" and report a bug
[430771.266732] nvme 0000:13:00.0: enabling device (0000 -> 0002)
[430771.266814] nvme nvme2: Disabling device after reset failure: -19
[430771.283392] I/O error, dev nvme2n1, sector 1812765936 op 0x0:(READ) flags 0x0 phys_seg 1 prio class 2
[430771.283397] zio pool=rpool vdev=/dev/nvme2n1p1 error=5 type=1 offset=928127770624 size=16384 flags=180880
[430771.283397] zio pool=rpool vdev=/dev/nvme2n1p1 error=5 type=1 offset=1394183585792 size=24576 flags=180880
[430771.283397] zio pool=rpool vdev=/dev/nvme2n1p1 error=5 type=2 offset=1575062740992 size=4096 flags=180880 [430771.283399] nvme2n1: detected capacity change from 3907029168 to 0
```
At this point, if I am lucky enough, I can manage to bring it back to life using a sledgehammer:
```
echo 1 > /sys/bus/pci/devices/0000\:12\:00.0/remove
echo 1 > /sys/bus/pci/rescan
```
If the faulted device reappears the zpool becomes ONLINE again as it completes its re-silvering (however the `zfs send` aborts obviously) else I am due to do a hardware reset or even a power cycle of the whole machine. 

I suspect a hardware issue: My best guess is that the Western Digital SN 770 modules controller is not not beefy enough to handle a burst of I/O requests so it is put on its knees and become so unresponsive that it is unable to complete a reset request on its own.  I also tried with the suggested `nvme_core.default_ps_max_latency_us=0 pcie_aspm=off` on the kernel command-line but no improvement and same symptoms so it *might not* be a power management issue, at least at first glance (unless something is really buggy there). As not always the same module crashes, they do not seems be all defective (or I am extremely unlucky).  Pool scrubbing might by a bit lighter for the controller so the scrub/resilver works without any issue (max speed is around 4.5~5 GB/s when scrubbing according to `zpool status`). 

I ran a short S.M.A.R.T. test: nothing special to say, all of the  three NMVE modules pass it. 

Had anybody here experienced something like this? Perhaps a workaround would be to try to throttle the number of IOPS my NVME devices handle? I am looking for hints/ideas.

PS: Neither  specific to ZFS 2.1.10 not Linux 6.2.x, I have had this issue with previous other versions as well for quite some time. I know using a not officially supported kernel "voids the warranty" but it does not seems to be a software issue.

**Comment by rincebrain** [comment]
This seems to be a hardware problem with your machine or an interaction with Linux, not ZFS specific.

You could have posted this as a discussion, but it does not seem to be a bug in ZFS at all.

You might also want to see if your PSU is beefy enough, and your wall power reliable enough, because multiple devices misbehaving sounds like either they're all faulty or something they have in common is, and I would suspect, if not an interaction of power management, either overheating, a firmware bug, or an interaction of actual power draw.

**Comment by admnd** [comment]
@rincebrain  your impression goes in the same direction than mine. The PSU is of well known brand and is not even that loaded (30% or so). I am not inclined to suspect brownouts at first glance (it is connected on the wall plug via a UPS that conditions the voltage and filter the voltage spikes). But who knows? The hardware can be "touchy" sometimes. I already had seen some servers of a very well established brand not tolerating SSDs from another very well established brand, why? Go figure even the technical support had no answer :)

So what you suggest is still plausible but I would have (probably) seen other symptoms (random system crash, memory errors and others weird symptoms), in all cases: good suggestion.

My objective was more to document the issue rather finding the happy path with a more or less magic solution in the middle. Hoping it brings some light in and even allows someone to make a wiser choice than mine on what to waste  the money on ;)  If there is a more suitable place to report/continue that discussion, I would be happy to put it there.

Thinking  more about it: those modules are designed for the general public Desktop PC usage. My use case here is certainly more the exception than the rule and it has not have been aggressively tested (or even tested at all) by the WD engineers themselves. Perhaps replacing those with WD Black SN850x (has a DRAM cache and a better controller)  or something equivalent would be more suitable for zpools in RAIDZ configurations. I might have some alternative NVMe modules somewhere here, this definitely worth a detour, else... will try to negotiate with the piggybank if...  by chance..... Anyway, I will put more information if I go that way. 

In all cases, thank you so much for having taken the time to go through my long initial post and put some suggestions. I appreciate it and I keep them under the hand.

On another subject non directly related to this one (and I need to underline it) in parenthesis.

( <-- parenthesis opened
To the gentleman who put a "thumb down" on the original post:
Option 1: it is an error (or you are a bot) and nothing have to be said, it just happens, end of the story.
Option 2: It is intentional from a human being (**_why?_**) and you are  entitled to my following personal opinion which is:

I have a life and in mine there are  tons of other higher personal priorities and various issues to sort out making my personal free time somewhat very scarce, especially this spring. I spent more than 2 hours to carefully search around, scrub Google results, study the issue in details, report the issue here and test some workaround stuff for the benefit of the community (read: **you** included) in the case of someone else would encounter the very same issue. I could have done many other things much more important from a personal vantage point during that time and kept the acquired knowledge just for me and let the others in the dark. After all I found nearly nothing around, so I was more less on my own.

I am not paid neither be under obligation of anything nor retire any direct personal benefit beyond having a reliable and well documented software. So returning some of the efforts to the community in a way or another (reporting and fixing bugs, writing documentation, helping others, etc) is my way to reward the community for all it provides so generously. True for software projects, true for anything else I kindly **choose** to put some **benevolent time and energy** on.   So the next time your are not happy with a bug report for a reason or another, do yourself a favor: Just ignore it.  Benevolent, good-wiling people are certainly not entitled to disgraceful reactions and **respect** is always to way to go. True for me true for anyone else on the planet who is involved in a cause that provides some public benefit in a way or another (developing software, participating in charitable organizations, taking care of others, etc).

Usually I (really) don't care about "likes" mentions and related, but it is a good opportunity to remind that personal altruism is the engine of many things especially in the free software world, ZFS making no exception. All of those tremendous benevolent efforts from the community  has no price and a "Thank you" is always appreciated. If have a personal issue with me just drop me note about this, I am generally opened to constructive discussions with smart guys, the rest being ignored, no time to waste with angry/frustrated people. Been there, done that. 

Like we say in french: _points mis sur les « i » et les barres sur les « T » passons à autre chose._

) <- parenthesis closed, no further discussion on this topic here.
